mcstatus/motd.py

```diff
@@ -0,0 +1,280 @@
+            raise ValueError("`hex` and `rgb` attributes is None, we need one of them.")
+
+        if self.hex is None:
+            if not all(256 > e > -1 for e in self.rgb):
+                raise ValueError("One of RGB colors is less than zero and more than 255: " + str(self.rgb))
+            self.hex = "#%02x%02x%02x" % self.rgb
+        else:
+            if not self.hex.startswith("#"):
+                self.hex = "#" + self.hex
+            hex = self.hex.lstrip("#")
+            try:
+                self.rgb = tuple(int(hex[i : i + 2], 16) for i in (0, 2, 4))  # type: ignore[assignment]
+            except ValueError:
+                raise ValueError(f"Incorrect hex color '{self.hex}', failed to generate RGB.")
+
+
+RawMotd = List[Union[Formatting, MinecraftColor, WebColor, str]]
+"""Type alias to field ``Motd.raw``."""
+
+
+@dataclass
+class Motd:
+    """Represents parsed MOTD.
+
+    Basing on thus class, you can write your own MOTD-to-something parser, see ``raw`` field.
+    """
+
+    raw: RawMotd
+    """Parsed MOTD, which then will be transformed.
+
+    Bases on this field, you can easily write your own MOTD-to-something parser.
+    """
+    bedrock: bool = False
+    """Is server Bedrock Edition? Some details may change in work of this class."""
+
+    @classmethod
+    def parse(cls, raw: Union[dict, list, str], *, bedrock=False) -> Self:
+        """Parse a raw MOTD to less raw MOTD (``Motd.raw`` field).
+
+        :param raw: Raw MOTD, directly from server.
+        :param bedrock: Is server Bedrock Edition? Nothing changes here, just sets attribute.
+        :returns: New ``MOTD`` instance.
+        """
+        if isinstance(raw, str):
+            return cls._parse_as_str(raw, bedrock=bedrock)
+
+        if isinstance(raw, list):
+            raw = {"extra": raw}
+
+        return cls(cls._parse_as_dict(raw), bedrock)
+
+    @staticmethod
+    def _parse_as_dict(item: dict, *, auto_add: Optional[RawMotd] = None) -> RawMotd:
+        raw_motd: RawMotd = auto_add if auto_add is not None else []
+
+        if item.get("color"):
+            try:
+                raw_motd.append(MinecraftColor[item["color"].upper()])
+            except KeyError:
+                raw_motd.append(WebColor(hex=item["color"]))
+
+        for style_key, style_val in Formatting.__members__.items():
+            if item.get(style_key.lower()):
+                raw_motd.append(style_val)
+
+        if (end := item.get("text")) is not None:
+            raw_motd.append(end)
+        if len(tuple(filter(lambda e: type(e) is Formatting, raw_motd))) > 0:
+            raw_motd.append(Formatting.RESET)
+
+        if item.get("extra"):
+            auto_add = list(filter(lambda e: type(e) is Formatting and e != Formatting.RESET, raw_motd))
+
+            parsed_extra: RawMotd = []
+            for element in item["extra"]:
+                parsed_extra.extend(Motd._parse_as_dict(element, auto_add=auto_add.copy()))
```

Comment:
I don't think mutating `auto_add` here is an issue, since it's not getting used anywhere below this line, we can save some memory by not needlessly copying it.

Reply:
If we will not copy it, next iteration will have `raw_motd` variable from previous iteration instead of real `auto_add`.

So example we have `auto_add = [Formatting.BOLD]` and `{"underlined": True, "text": "hi"}`. After first iteration our `auto_add` variable will be `[Formatting.BOLD, Formatting.UNDERLINED, "hi", Formatting.RESET]` and on second iteration we will give new `auto_add` list instead of actual one.